llama.cpp

```diff
@@ -959,6 +959,13 @@ struct llama_vocab {
 
     id linefeed_id = 13;
 
+    // codellama FIM special tokens
+    // TODO: load these from the vocabulary.
+    id special_prefix_id = 32007;
+    id special_middle_id = 32009;
+    id special_suffix_id = 32008;
+    id special_eot_id = 32010;
```

Comment:
Curious, does Code Llama 34B have these special tokens?
If it does not, then how would FIM work with it?

Reply:
It does, yeah. These are new, and I think only in codellama. I don't think they're in llama2. To get the token ids themselves, the codellama people run the tokenizer, and these are the values that came out.

https://github.com/facebookresearch/codellama/blob/cb51c14ec761370ba2e2bc351374a79265d0465e/llama/tokenizer.py#L28-L31

It should work. But I've been busy with my day job, and haven't gotten a chance to test it yet. Definitely not going to suggest merging until I'm certain.